src/Background/Portal.vala

```diff
@@ -0,0 +1,171 @@
+            warning ("Cannot connect to compositor, background portal working with reduced functionality.");
+        }
+    }
+
+    [DBus (name = "org.pantheon.gala.DesktopIntegration")]
+    public interface DesktopIntegration : Object {
+        public struct RunningApplications {
+            string app_id;
+            HashTable<string,Variant> details;
+        }
+
+        public signal void running_applications_changed ();
+        public abstract RunningApplications[] get_running_applications () throws DBusError, IOError;
+    }
+
+    private enum ApplicationState {
+        BACKGROUND,
+        RUNNING,
+        ACTIVE
+    }
+
+    public HashTable<string, Variant> get_app_state () throws DBusError, IOError {
+        if (desktop_integration == null) {
+            throw new DBusError.FAILED ("No connection to compositor.");
+        }
+
+        var apps = desktop_integration.get_running_applications ();
+        var results = new HashTable<string, Variant> (null, null);
+        foreach (var app in apps) {
+            var app_id = app.app_id;
+            if (app_id.has_suffix (".desktop")) {
+                app_id = app_id.slice (0, app_id.last_index_of_char ('.'));
+            }
+
+            results[app_id] = ApplicationState.RUNNING; //FIXME: Don't hardcode: needs implementation on the gala side
+        }
+
+        return results;
+    }
+
+    public async void notify_background (
+        ObjectPath handle,
+        string app_id,
+        string name,
+        out uint32 response,
+        out HashTable<string, Variant> results
+    ) throws DBusError, IOError {
+        uint32 _response = 2;
+        var _results = new HashTable<string, Variant> (str_hash, str_equal);
+
+        var notification_request = new NotificationRequest ();
+
+        notification_request.response.connect ((result) => {
+            switch (result) {
+                case NotificationRequest.NotifyBackgroundResult.CANCELLED:
+                    _response = 1;
+                    break;
+                case NotificationRequest.NotifyBackgroundResult.FAILED:
+                    break;
+                default:
+                    _response = 0;
+                    _results["result"] = result;
+                    break;
+            }
+
+            notify_background.callback ();
+        });
+
+        uint register_id = 0;
+        try {
+            register_id = connection.register_object<NotificationRequest> (handle, notification_request);
+        } catch (Error e) {
+            warning ("Failed to export request object: %s", e.message);
+            throw new DBusError.OBJECT_PATH_IN_USE (e.message);
+        }
+
+        notification_request.send_notification (app_id, name);
+
+        yield;
+
+        connection.unregister_object (register_id);
+        response = _response;
+        results = _results;
+    }
+
+    private enum AutostartFlags {
+        AUTOSTART_FLAGS_NONE,
+        AUTOSTART_FLAGS_DBUS_ACTIVATABLE
+    }
```

Comment:
```suggestion
    private enum AutostartFlags {
        NONE,
        DBUS_ACTIVATABLE
    }
```

Reply:
Ah right I copied that from the gtk implementation 😅